packages/grid/_modules_/grid/components/columnSelection/GridHeaderCheckbox.tsx

```diff
@@ -28,7 +35,10 @@ export const GridHeaderCheckbox = React.forwardRef<HTMLInputElement, GridColumnH
 
     const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
       const checked = event.target.checked;
-      apiRef!.current.selectRows(visibleRowIds, checked, !event.target.indeterminate);
+      const rowsToBeSelected = checkboxSelectionVisibleOnly
+        ? paginatedVisibleRowIds
```

Comment:
We are running both selectors even when the callback is not called
Is it totally negligible performance-wise or should we add a callback version of the selectors ?
Something like :

```ts
const getVisibleRowIds = useGridSelectorCallback(apiRef, visibleSortedGridRowIdsSelector)
const getPaginatedVisibleRowIds = useGridSelectorCallback(apiRef, gridPaginatedVisibleSortedGridRowIdsSelector)

...

const handleChange = () => {
  const rowsToBeSelector = checkboxSelectionVisibleOnly
    ? getPaginatedVisibleRowIds()
    : getVisibleRowIds()
}
```

And even without it, some selectors are performing actions like filter on large table, but from what I understand, a selector is never memoized, it run at each render of the component using `useGridSelector`.
But this is not exactly related to this PR.

```ts
export const useGridSelector = <State>(
  apiRef: GridApiRef | undefined,
  selector: (state: any) => State,
) => {
  const [state] = useGridState(apiRef!);
  return selector(state);
};
```

Reply:
we memoize at the selector level using `createSelector` from 'reselect'
https://github.com/reduxjs/reselect#creating-a-memoized-selector

Reply:
Oh indeed, my bad.
So forget my second point.

And for the 1st one, I suppose you consider that selectors like `visibleSortedGridRowIdsSelector` will always be memoized somewhere by another component that does need it on every render ?

Reply:
If it's a selector it will be memoized and recomposed at the first call when the state change. It doesn't matter if it is called 1 or 100 times.

Reply:
@flaviendelangle 👍 for calling the selector directly in `handleChange()` with the state from `useGridState`. We don't need to run the slice each time the visible rows change. In this very instance, it probably doesn't really matter, but I think that it could set an interesting precedent for this more systematized "lazy" pattern. 

The rule could be:

- `useGridSelector` if the value is used in the render method.
- call the selector directly if used inside a handler.

Redux **nor** Recoil supports this pattern, why? what's the downside?

Reply:
I can see a down side when we are using the callback in an effect and want it to run whenever the dependencies of the callback are changing.

Reply:
This discussion seems to go off-topic from that is in the PR. We can always move it to a dedicated discussion.

I've updated the PR to call the selectors directly inside the `handleChange`.